sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -617,6 +618,77 @@ object NullPropagation extends Rule[LogicalPlan] {
 }
 
 /**
+ * Propagate foldable expressions:
+ * Replace all attributes with aliases of the original foldable expressions except the followings.
+ * 1) Command and Set(UNION/INTERSECT/EXCEPT): Do not optimize.
+ * 2) Filter/Sort: Use the original foldable expressions without aliases.
```

Comment:
That's a good point. I'll try to mix `CleanupAlias` and `FoldablePropagation` and simplify `FoldablePropagation`.